### ISSUE TITLE
Publish the cross-compile toolchain as a container image

### ISSUE THREAD

**Issue by qrkourier** [openziti/ziti-tunnel-sdk-c#646]
The purpose of this change is to improve the developer experience for T-SDK, specifically building the `ziti-edge-tunnel` (ZET) binary. Building the C-SDK and ZET packages (RPM, DEB) are not in scope for this issue.

The new container image will be versioned separately from T-SDK and live in a separate repo. It will be based on [the unofficial cross-build container](https://github.com/openziti/ziti-tunnel-sdk-c/blob/main/docker/Dockerfile.linux-cross-build), which will be updated to use Ubuntu Bionic for GLIBC 2.27.

Then, a version of the new container image will be pinned to the CMake workflows for the T-SDK, replacing the current job-container, `ubuntu:18.04` ([link to workflow](https://github.com/openziti/ziti-tunnel-sdk-c/blob/main/.github/workflows/cmake.yml#L41)).

**Comment by qrkourier** [comment]
Please comment on whether this implementation meets your expectations for a cross-compile container image: https://github.com/openziti/ziti-tunnel-sdk-c-builder#readme.

**Comment by qrkourier** [comment]
@ekoby has suggested that we broaden the scope of the container image to be a more general-purpose build tool shared by the CI and local build scripts in the following projects:

* T-SDK
* C-SDK
* zitify

I am beginning to investigate the build requirements for C-SDK and zitify since they may be different from T-SDK. I propose the follow developer experience for local builds for each project:

A build.sh script in the repo will have the default behavior of emulating the CMake workflow in CI. Optionally, the developer may override the default build target, etc. by specifying arbitrary CMake options and arguments. 

As a stretch goal, the build script could detect the host arch of the developer's workstation and automatically compile for that target arch. Please comment on whether this would meet your expectations.

Notably, I do not plan to design a CLI/API for the build procedure, i.e. replete opts and args for the developer to learn in order to interact with the build script. If this is an important goal for you then please comment on how you want that to work and why.

**Comment by qrkourier** [comment]
This solution will replace some of the steps in [this repo's CMake CI workflow](https://github.com/openziti/ziti-tunnel-sdk-c/blob/main/.github/workflows/cmake.yml#L109). For example, the steps that install and run `cmake` and `vcpkg`. If the same approach is adopted by the C-SDK or zitify repos then the same types of steps in those repos' CMake workflows will be replaced by a single step to run this container.

**Comment by ekoby** [comment]
I would prefer that Docker image does not make assumptions about the build process or scripts. For example in CI, I just want to use this image like this
```
jobs:
  build:
    runs-on: ubuntu-latest
    container:
      image: ziti-c-builder:latest
```
so I can still use github actions that provide useful utilities, like caching

**Comment by qrkourier** [comment]
That makes sense. I'd like to propose the following solution summarizing the input I've received up to this point. Please comment on whether this meets your expectations for a CMake container image.

* a published builder container image `openziti/ziti-cmake` with `WORKDIR /github/workspace` for GH Actions compatibility
* image has on-board all the necessary utilities and libraries for cross-compiling OpenZiti projects with CMake (those that employ the C-SDK)
* image does not run `cmake` or produce any particular artifact because it has no entrypoint script or default command except `bash`
* each project will have a new shell script `./build.sh` that is intended to be run _inside_ the CMake container to build the default target for that project
* `./build.sh` will always delete `/github/workspace/build` before running `cmake` with build output dir `./build`
* the GitHub Actions CMake workflow for each project will use the `ziti-cmake` container as a job container in which to execute job steps, including a `run` step to run `./build.sh`
* developers will run `./build.sh` locally when they wish to invoke the same build procedure expressed in the GitHub Actions CMake workflow and they will have the choice to run the script on their workstation's filesystem or inside the CMake container with the project mounted on `/github/workspace`

**Comment by qrkourier** [comment]
* developers may enhance the `./build.sh` in each project with options and arguments to conveniently build different targets, etc.

**Comment by qrkourier** [comment]
I have updated [the README.md](https://github.com/openziti/ziti-cmake/pull/1/files?short_path=b335630#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5) in this issue's resolving pull request to reflect the implementation sketch above.

**Comment by qrkourier** [comment]
> the prescription for cmake.sh seem superfluous in this context, the consumer project can do whatever, right?

That's true. It's technically superfluous because the container image will not prescribe any command or use case, assuming the premeditated series of changes in the repositories that will consume this new container image are acceptable as described here. That assumption is probably correct. If that assumption is incorrect, and the consuming repos _must_ use the image differently for some reason, then it will change the design of this image.

**Comment by qrkourier** [comment]
Here's an implementation of `cmake.sh` using this container image to cross-compile the Linux artifacts in CI. A developer can run `./cmake.sh` to build the default target for x86_64 locally, or `./cmake.sh -p ci-linux-arm64` to build the default target with an alternative os-architecture preset.

https://github.com/openziti/ziti-tunnel-sdk-c/pull/662

**Comment by qrkourier** [comment]
```bash
❯ ./cmake.sh help

Usage: cmake.sh [CMD] [ARGS...] 

Runs CMD in the ziti-cmake container, and builds the 
default target if no CMD is specified
 
    -c  CMAKE_CONFIG  set CMAKE_BUILD_TYPE (default: Release) 
    -p  CMAKE_PRESET  set CMAKE_TOOLCHAIN_FILE preset (default: ci-linux-x64) 
    -t  CMAKE_TARGET  set CMAKE_TARGET (default: bundle)
```
